services/brig/test/integration/API/OAuth.hs

```diff
@@ -0,0 +1,865 @@
+    let refreshAccessTokenRequest = OAuthRefreshAccessTokenRequest OAuthGrantTypeRefreshToken cid secret (oatRefreshToken accessToken)
+    threadDelay $ 2 * 1010 * 1000 -- wait for 2 seconds for the token to expire
+    refreshOAuthAccessToken' brig refreshAccessTokenRequest !!! do
+      const 403 === statusCode
+      const "Forbidden" === statusMessage
+
+testRefreshTokenRevokedToken :: Brig -> Http ()
+testRefreshTokenRevokedToken brig = do
+  uid <- userId <$> createUser "alice" brig
+  let redirectUrl = mkUrl "https://example.com"
+  let scopes = OAuthScopes $ Set.fromList [ReadSelf]
+  (cid, secret, code) <- generateOAuthClientAndAuthorizationCode brig uid scopes redirectUrl
+  let accessTokenRequest = OAuthAccessTokenRequest OAuthGrantTypeAuthorizationCode cid secret code redirectUrl
+  accessToken <- createOAuthAccessToken brig accessTokenRequest
+  let refreshAccessTokenRequest = OAuthRefreshAccessTokenRequest OAuthGrantTypeRefreshToken cid secret (oatRefreshToken accessToken)
+  revokeOAuthRefreshToken brig (OAuthRevokeRefreshTokenRequest cid secret (oatRefreshToken accessToken)) !!! const 200 === statusCode
+  refreshOAuthAccessToken' brig refreshAccessTokenRequest !!! do
+    const 403 === statusCode
+    const "Forbidden" === statusMessage
+
+testListApplicationsWithAccountAccess :: Brig -> Http ()
+testListApplicationsWithAccountAccess brig = do
+  uid <- userId <$> createUser "alice" brig
+  do
+    apps <- listOAuthApplications brig uid
+    liftIO $ assertEqual "apps" 0 (length apps)
+  void $ createOAuthApplicationWithAccountAccess brig uid
+  void $ createOAuthApplicationWithAccountAccess brig uid
+  do
+    apps <- listOAuthApplications brig uid
+    liftIO $ assertEqual "apps" 2 (length apps)
+  void $ createOAuthApplicationWithAccountAccess brig uid
+  do
+    apps <- listOAuthApplications brig uid
+    liftIO $ assertEqual "apps" 3 (length apps)
+
+testRevokeApplicationAccountAccess :: Brig -> Http ()
+testRevokeApplicationAccountAccess brig = do
+  uid <- userId <$> createUser "alice" brig
+  do
+    apps <- listOAuthApplications brig uid
+    liftIO $ assertEqual "apps" 0 (length apps)
+  for_ [1 .. 3 :: Int] $ const $ createOAuthApplicationWithAccountAccess brig uid
+  cids <- fmap oaId <$> listOAuthApplications brig uid
+  liftIO $ assertEqual "apps" 3 (length cids)
+  case cids of
+    [cid1, cid2, cid3] -> do
+      revokeOAuthApplicationAccess brig uid cid1
+      do
+        apps <- listOAuthApplications brig uid
+        liftIO $ assertEqual "apps" 2 (length apps)
+      revokeOAuthApplicationAccess brig uid cid2
+      do
+        apps <- listOAuthApplications brig uid
+        liftIO $ assertEqual "apps" 1 (length apps)
+      revokeOAuthApplicationAccess brig uid cid3
+      do
+        apps <- listOAuthApplications brig uid
+        liftIO $ assertEqual "apps" 0 (length apps)
+    _ -> liftIO $ assertFailure "unexpected number of apps"
+
+testWriteConversationsSuccessInternal :: Brig -> Nginz -> Http ()
```

Comment:
up to here, the order was the same in the test groups and the test definitions.  that was nice.  :)

Reply:
also the 3 `...Internal` functions are not used anywhere, plus they are exact copies of the `Nginz` functions.  remove, i think?